rls/src/main/java/io/grpc/rls/internal/LbPolicyConfiguration.java

```diff
@@ -0,0 +1,369 @@
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("effectiveRawChildPolicy", effectiveRawChildPolicy)
+          .add("effectiveLbProvider", effectiveLbProvider)
+          .add("childPolicyConfigTargetFieldName", targetFieldName)
+          .toString();
+    }
+  }
+
+  /** Factory for {@link ChildPolicyWrapper}. */
+  static final class RefCountedChildPolicyWrapperFactory {
+    @VisibleForTesting
+    final Map<String /* target */, RefCountedChildPolicyWrapper> childPolicyMap =
+        new HashMap<>();
+
+    private final ChildLoadBalancerHelperProvider childLbHelperProvider;
+    @Nullable
+    private final ChildLbStatusListener childLbStatusListener;
+
+    public RefCountedChildPolicyWrapperFactory(
+        ChildLoadBalancerHelperProvider childLbHelperProvider,
+        @Nullable ChildLbStatusListener childLbStatusListener) {
+      this.childLbHelperProvider = checkNotNull(childLbHelperProvider, "childLbHelperProvider");
+      this.childLbStatusListener = childLbStatusListener;
+    }
+
+    ChildPolicyWrapper createOrGet(String target) {
+      // TODO(creamsoup) check if the target is valid or not
+      ObjectPool<ChildPolicyWrapper> existing = childPolicyMap.get(target);
+      if (existing != null) {
+        return existing.getObject();
+      }
+      ChildPolicyWrapper childPolicyWrapper =
+          new ChildPolicyWrapper(target, childLbHelperProvider, childLbStatusListener);
+      RefCountedChildPolicyWrapper wrapper = RefCountedChildPolicyWrapper.of(childPolicyWrapper);
+      childPolicyMap.put(target, wrapper);
+      return childPolicyWrapper;
+    }
+
+    void release(ChildPolicyWrapper childPolicyWrapper) {
+      checkNotNull(childPolicyWrapper, "childPolicyWrapper");
+      String target = childPolicyWrapper.getTarget();
+      ObjectPool<ChildPolicyWrapper> existing = childPolicyMap.get(target);
+      checkState(existing != null, "Cannot access already released object");
+      if (existing.returnObject(childPolicyWrapper) == null) {
+        childPolicyMap.remove(target);
+      }
+    }
+  }
+
+  /**
+   * ChildPolicyWrapper is a wrapper class for child load balancing policy with associated helper /
+   * utility classes to manage the child policy.
+   */
+  static final class ChildPolicyWrapper {
+
+    private final String target;
+    private final ChildPolicyReportingHelper helper;
+    private ConnectivityStateInfo connectivityStateInfo =
+        ConnectivityStateInfo.forNonError(ConnectivityState.IDLE);
+    private SubchannelPicker picker;
+
+    public ChildPolicyWrapper(
+        String target,
+        ChildLoadBalancerHelperProvider childLbHelperProvider,
+        @Nullable  ChildLbStatusListener childLbStatusListener) {
+      this.target = target;
+      this.helper =
+          new ChildPolicyReportingHelper(childLbHelperProvider, this, childLbStatusListener);
```

Comment:
This circular dependency looks weird, because `ChildPolicyWrapper` is consumed by `new ChildPolicyReportingHelper()` before instantiation is complete.

Making `ChildPolicyReportingHelperImpl` an inner (non-static) class of `ChildPolicyWrapper` might be more natural.

Reply:
yes i was wondering same thing.